Add NSSL 2/3-moment cloud microphysics

MicroTed opened the pull request:
Add NSSL 2/3-moment cloud microphysics

Adds option to use the NSSL 2-moment cloud microphysics scheme with 6 hydrometeor species (droplets, rain, cloud ice, snow, graupel, and hail). It includes the option for 3-moment rain/graupel/hail (adds reflectivity moment) and an option to turn off hail (for non-severe convection). Details on using the scheme are in docs/README.NSSLmp. Also adds a 3D reflectivity field for optional output (refl10cm). The parameterization code is (or will be) at parity with the current WRF PR for the 3-moment scheme https://github.com/wrf-model/WRF/pull/1876 (and eventual CCPP update).

Tested on daily 3-km regional CONUS forecasts (with v7.3), so similar performance is expected with v8.0

Outstanding issue: At least in the physics_mmm directory, the .mod files are not updated by recompiling newer .F files. They have to be deleted in order to be replaced. I don't know if this is general or a gfortran quirk.

MicroTed committed:
Initial implementation of NSSL 2-moment microphysics

MicroTed committed:
Added a package (nssl_moments) to turn on 3-moment version of NSSL microphysics (adds refl. moment variables for rain, graupel, hail: zrw, zgw, zhw).
 Also added code to compute the composite and 1-km reflectivity for NSSL2M (not actually tested yet)

MicroTed committed:
Limit ice crystal effective radius to 130um to stay within allowed RRTMG range.

MicroTed committed:
Enable substepping for mp_nssl2m when dt_dyn is large (automatically sets n_microp)

MicroTed committed:
Added new variables to LBC code in time_integration

MicroTed committed:
Add support for compression for netcdf4
  mpas_io_streams.F : pass ioformat to def_var
  mpas_io.F : add conditional to set deflate on variables if ioformat is netcdf4 (needs netcdf-4.7.4 or later)

MicroTed committed:
Hack to set chunksizes along the nCells dimension. Chunksize is nCells/nprocs

MicroTed committed:
Added check in Makefile for HDF5 env. variable, which seems to be needed if the libraries are static only.

MicroTed committed:
Fixed out-of-bounds problem with dimnames array

MicroTed committed:
Add support for 3D reflectivity output (refl10cm) from Thompson scheme

MicroTed committed:
Merge branch 'nssl2m-deflate' of https://github.com/MicroTed/MPAS-Model into nssl2m-deflate

MicroTed committed:
Added mp_nssl2m to schemes that support calculated effective radii.

MicroTed committed:
Merge branch 'release-v8.0.0' into nssl2m-deflate

MicroTed committed:
Registry.xml : Add variables/packages and control flags for NSSL microphysics
ccpp_kinds.F : Add 'machine' module for CCPP compatibility
mpas_atmphys_driver_microphysics.F : Add support for NSSL microphysics (arrays, init/driver etc.)
mpas_atmphys_interface.F : Get/put data for NSSL microphysics
mpas_atmphys_packages.F : Set up packages for NSSL microphysics
physics_mmm/Makefile : add dependencies for NSSL-mp
physics_wrf/Makefile : remove NSSL-mp (moved to physics_mmm)
docs/README.NSSLmp : brief overview of options in the NSSL microphysics scheme
physics_mmm: module_mp_nssl_2mom.F, mp_nssl.F : new code file for NSSL-MP

MicroTed committed:
Merge remote-tracking branch 'origin/hotfix-v8.0.1' into v801-nssl3m

MicroTed committed:
Merge branch 'develop' into dev-nssl3m

MicroTed committed:
Added surface accumulation/rate for hail (separate from graupel)

MicroTed commented:
Hi, @ldfowler58 would you have time to look at this at some point? Let me know if this is the approach you want for physics_mmm or if there is more to shift from the model code to the interface. The physics module is an updated version from CCPP, so it has all the extra comments for documentation. The interface 'run' subroutine is from CCPP (for SCM/FV3), but isn't actually used since MPAS can call the microphysics directly with the "_p" arrays.
Thanks!

mgduda commented:
I think it would be good to separate the changes for NetCDF compression from the changes to add the NSSL microphysics scheme, as these (if I understand correctly) are unrelated changes.

MicroTed commented:
I agree that it would be cleaner to split it off the netcdf part, and it is only 2 files and I think no overlapping changes. It's a bare-bones implementation that assumes one would only choose the netcdf4 option because of compression being available. Otherwise the speed of pnetcdf for uncompressed writes really can't be beat.

MicroTed committed:
Reverted 2 files that had changes to enable variable compression with the netcdf4 output option

MicroTed commented:
I moved the netcdf changes to a separate branch:  https://github.com/MicroTed/MPAS-Model/tree/dev-nc4deflate
I could make a draft PR or just leave it on ice for now -- whatever you want to do given your priorities and limited time. Ideally I would add a netcdf version check and an extra flag for deflate level etc.

ldfowler58 commented:
I started to look at the changes and add-ons made by Ted in order to include the NSSL cloud microphysics scheme. Here my first suggestions for the top files that I have been modified:

1) In ./src/core_atmosphere/Registry.xml, I suggest to keep the "mp_" reference in front of the packages named nssl_hail_in, nssl_ccn_in, nssl3m_in, and nssl3m_hail_in. Users may appreciate that those packages point to cloud microphysics options. nssl3m_in can be changed to mp_nssl3m_in since you defined the NSSL 2-moment package as mp_nssl2m_in. Thanks.

2) In ./src/core_atmosphere/physics/mpas_atmphys_packages.F, can you please keep everything "lower case" and remove extra blank spaces to ensure "continuity" in how the sourcecode has been originally written (if not much to ask). For instance, change "
    if(config_nssl_hail_on) then
      nssl_hail_in = .true.
   endif

3) In ./src/core_atmosphere/physics/mpas_atmphys_control.F, change lines 
           config_microp_scheme .eq. 'mp_thompson' .or. &
           config_microp_scheme .eq. 'mp_nssl2m'  .or. &
to
           config_microp_scheme .eq. 'mp_nssl2m'      .or. &
           config_microp_scheme .eq. 'mp_thompson' .or. &

4) In ./src/core_atmosphere/physics/mpas_atmphys_vars.F (lines 197-203), do you mind putting one variable per line as for lines 188-195 so that I can add the definition of the different variables later. Thanks.


5) In ./src/core_atmosphere/physics/mpas_atmphys_manager.F, same comments as in 2) regarding lower case and space. Wouldn't it be simpler to use 75._RKIND instead of 1.25001*60._RKIND? Thanks.

Will look at the actual microphysics next. Thanks.

MicroTed commented:
Thanks for the comments! My thinking for leaving off "mp_" from nssl3m_in and nssl_hail_in etc. is that they are package options rather than full packages. They are not seen by the user, but rather activated by config_ options. So maybe it doesn't really matter? The package is still the mp_nssl2m_in, and the others are options (subpackages?) to turn variables on or off. 

Perhaps I could just rename mp_nssl2m_in to just mp_nssl_in? Or something like that? I also have a version with an option to run single-moment, but I'm not sure that's needed for a release code.

I can change the other formatting things to match the rest of the code. (Although I'm firmly in the camp of having a space after 'if' :) considering there is an unbalanced space in front of 'then', haha.)

MicroTed committed:
Formatting updates

MicroTed committed:
Merge remote-tracking branch 'origin/develop' into dev-nssl3m

MicroTed committed:
mpas_atm_time_integration.F : add lines to get indices for added scalars (qh, ng, nh, etc.)

MicroTed committed:
Merge remote-tracking branch 'origin/hotfix-v8.2.1' into dev-nssl3m

MicroTed committed:
In core_atmosphere:
  Registry.xml : fix ntiedtke vs. tiedtke
  physics/ccpp_kind_types.F : remove 'machine'
  physics/mpas_atmphys_interface.F : fix typo and error with refl10cm
  physics/physics_wrf/Makefile : add NSSL mp routines (for now)
  physics/physics_wrf/mp_nssl.F : switched module for kind_phys

MicroTed committed:
Merge branch 'dev-nssl3m' of https://github.com/MicroTed/MPAS-Model into dev-nssl3m

MicroTed committed:
mpas_atmphys_control.F : fixed whitespace

MicroTed committed:
ccpp_kind_types.F : removed extra line break

MicroTed commented:
Updated (merged) to 8.2.1-hotfix. Also added function in init_atmosphere to generate LBCs for hydrometeor number concentrations when the LBC data only have mixing ratio. I still need to run some more tests and update the NSSL microphysics code. (Also some enhancements to the ideal initialization will be added.)

MicroTed committed:
Registry.xml : Removed nssl2m (for now) from tendencies; some logical reordering of scalars
  mpas_atmphys_interface.F : For mp_nssl, changes to avoid problems with PG compilers and pointers
  module_mp_nssl_2mom.F : Updates to recent development version
     - Use more accurate form for water saturation
     - Add active breakup parameterization for 3-moment rain
     - Fixes problem with sedimentation when nz > 128
     - Cleans up some unused/unset rate arrays that could cause issues in debug compile

MicroTed commented:
I ran into an odd compiler issue with nvfortran/pgf90 at -O3 with mpas_atmphys_interface.F: If a pointer assignment is inside an IF structure, it may not work correctly. For example, the optional hail variables here:

```
             IF ( config_nssl_hail_on ) THEN
              call mpas_pool_get_dimension(state,'index_qh'  ,index_qh  )
              call mpas_pool_get_dimension(state,'index_nh'  ,index_nh  )
              call mpas_pool_get_dimension(state,'index_volh' ,index_volh  )
               qh   => scalars(index_qh,:,:)
               nh   => scalars(index_nh,:,:)
               volh => scalars(index_volh,:,:)
             ENDIF
```

At -O0 it works fine, but -O2 or 3, the pointer for qh seems to go randomly somewhere, maybe within the scalar structure or maybe not. I have only looked at the min/max values of qh. Obviously this is bad compiler behavior, but I found two work-arounds for it:
 1) Remove the IF clause and just always assign qh etc. even if the index is -1, but still have an IF clause around accessing the array. I don't know if this will cause debug problems, however (e.g., bounds check, although the array isn't being accessed)
 2) Just replace 
                  qh_p(i,k,j)   = qh(k,i)
      with
                 qh_p(i,k,j)   = scalars(index_qh,k,i)

I currently have option 2 in the PR code, but can switch it to 1 or a better suggestion.

Also for this reason I have not yet included mp_nssl with the mass/number tendencies (e.g., rqcmpten). Those seem to be diagnostic only, but maybe there will be a need for those.

ldfowler58 commented:
Hi Ted: Thanks for pointing this out. I do not use nvfortran or pgf90 at all but it is definitely something that we should look into. Is it a "feature" that only occurs with the NSSL cloud microphysics or a "feature" that could also occur with, for instance, the Thompson cloud microphysics scheme? I'll investigate with a different cloud microphysics scheme separately.
Laura

MicroTed commented:
Hi, Laura,
It seems to only be a compiler bug with the IF-ENDIF sections, not with the SELECT-CASE at all. So I think the mp-thompson sections are OK. I did test the regular mp_thompson and it ran fine. We have some AMD epyc computers locally, and the nvfortran/pgf runs pretty well for them. I haven't tried gfortran yet for comparison, but will try to do that.

That said, I don't yet fully understand the structure of the 'scalar' array and how it works to reference the first index. (Basically how it converts a s(numscalar)%vals2d(nlevs,ncells) kind of record into something that looks like a 3D array -- too fancy!)

weiwangncar commented:
@MicroTed Apparently not updating .mod file is a gfortran feature as I just learnt. See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=31587

MicroTed committed:
mpas_atmphys_driver_microphysics.F: Pass value for f_cn to NSSL microphysics for case of ccn_on=F

MicroTed committed:
module_mp_nssl_2mom.F: set default droplet nucleation to option 5 (instead of 2), as this more realistically limits supersaturation.

MicroTed committed:
module_mp_nssl_2mom.F : Reduced max hail diam; Relaxed high Vt warning

MicroTed committed:
mpas_atmphys_packages.F : add write statements for NSSL mp config variables (when selected)
 module_mp_nssl_2mom.F : remove small noise values of hydrometeors at beginning of NSSL micro (already done at end)

MicroTed committed:
module_mp_nssl_2mom.F : Fixed out of bounds bug in the sedimentation

MicroTed committed:
Fixed default value of config_nssl_cccn; Fixed issue when CCN prediction is turned off, where the max droplet concentration could only be half of nssl_cccn.

MicroTed committed:
Reverted auto setting of ihlcnh for 2-moment that was meant for testing

MicroTed committed:
core_atmosphere/Registry.xml : Fixed default value of nssl_cccn

MicroTed committed:
Merge branch 'dev-nssl3m' of https://github.com/MicroTed/MPAS-Model into dev-nssl3m

MicroTed committed:
Reset change in Registry.xml

MicroTed committed:
Merge remote-tracking branch 'upstream/hotfix-v8.2.3' into v8.2.3-nssl

MicroTed committed:
module_mp_nssl_2mom.F : Update for better reflectivity conservation in graupel->hail conversion and for drop freezing.

MicroTed committed:
Merge remote-tracking branch 'origin/develop' into dev-nssl3m

MicroTed committed:
module_mp_nssl_2mom.F: Update a comment

MicroTed committed:
Merge remote-tracking branch 'upstream/develop' into dev-nssl3m

MicroTed committed:
A couple fixes for the refl10cm array

MicroTed committed:
core_atmosphere/Registry.xml : add LBC arrays for reflectivity moments
  mpas_init_atm_cases.F : add diagnosis of number concentrations

MicroTed committed:
Merge remote-tracking branch 'upstream/develop' into dev-nssl3m

MicroTed committed:
NSSL microphysics:
  - Moved code from physics_wrf to physics_noaa/NSSL-MP using manage externals to retrieve code from github
  - Added diagnostic code to mpas_init_atm_cases.F for reflectivity moment boundary conditions

MicroTed commented:
Updated to use the manage externals functionality to clone the microphysics code to physics_noaa

MicroTed committed:
Fixed case error in src/core_atmosphere/physics/Makefile

MicroTed commented:
Question for anybody: How much work would be involved to get the namelist filename from within mpas_atmphys_driver_microphysics.F?  (In the case where somebody is using the '-n' command line option to change it from the default.) The init subroutine takes the file name as an input, and it is currently hardwired to the default in the call. The reason is so that it can check for its internal namelist with rarely-used options. Most users would never use it, but it would be nice to correct this at some point.

EDIT: code has been updated to pass the namelist filename to the nssl init routine

MicroTed committed:
Update Externals.cfg to point to minor update of NSSL-MP

MicroTed committed:
Merge remote-tracking branch 'upstream/develop' into dev-nssl3m

jihyeonjang commented:
Hi Ted, 

Here are my suggestions based on my tests (before dev-nssl3m was merged into the current develop at [840ef2d](https://github.com/MPAS-Dev/MPAS-Model/pull/1113/commits/840ef2dc709065cd86c16d9c6bd15df2771d83d4)). 

 1. Radar reflectivity for Thompson scheme
There was an issue with the radar reflectivity output for the Thompson scheme after merging the NSSL-MP changes. To address this, please consider the following changes: 
 
**src/core_atmosphere/physics/mpas_atmphys_interface.F**
- Remove L1022 (no longer needed; related to the PR [#1329](https://github.com/MPAS-Dev/MPAS-Model/pull/1329)) 
```
L1022:        call mpas_pool_get_array(diag_physics,'refl10cm',refl10cm)
```
- Remove L1044 and L1053 for the Thompson scheme: 
```
L1044:              call mpas_pool_get_array(diag_physics,'refl10cm',refl10cm)
L1053:                 refl10cm(k,i) = refl10cm_p(i,k,j)
```

**src/core_atmosphere/physics/mpas_atmphys_driver_microphysics.F**
- Remove L163 and L274 for the Thompson scheme:
```
L163:           if(.not.allocated(refl10cm_p)) allocate(refl10cm_p(ims:ime,kms:kme,jms:jme))
L274:          if(allocated(refl10cm_p)) deallocate(refl10cm_p)
```

Note: The physics-intermediate refl10cm_p array was removed in the PR [#1329](https://github.com/MPAS-Dev/MPAS-Model/pull/1329), but it is still needed for NSSL-MP, since 10-cm radar reflectivity is computed in nssl_2mom_driver. 

&nbsp;

2. In src/framework/mpas_dmpar.F, the following lines are no longer needed after the commit [a5fc16f](https://github.com/MPAS-Dev/MPAS-Model/pull/1113/commits/a5fc16fbf5b42f09355b747982375a475fa31a1d). 
```
L87:   integer, public :: numprocs = 1
L313:       numprocs = mpi_size
```

&nbsp;

3. In src/core_atmosphere/physics/mpas_atmphys_packages.F, remove the following duplicated line. 
```
L127:    call mpas_log_write('    config_nssl_hail_on            = $l', logicArgs=(/config_nssl_hail_on/))
```

jihyeonjang commented:
I would also like to confirm the preferred namelist name for NSSL-MP: "mp_nssl2m" vs "mp_nssl". When config_nssl_3moment = .true., the 3-moment option becomes available, but still the microphysics name remains mp_nssl2m. Also, your branch name for this PR is dev-nssl3m. However, the subroutine names (nssl_2mom_init_const, nssl_2mom_driver, nssl_2mom_gs) are consistent with "mp_nssl2m", so mp_nssl2m could be more consistent. 

Could you decide which namelist name you prefer for NSSL-MP?

MicroTed commented:
> I would also like to confirm the preferred namelist name for NSSL-MP: "mp_nssl2m" vs "mp_nssl". When 

I agree it is awkward and was I just keeping the '2m' for legacy. It would make sense to simplify it to 'mp_nssl'. It can even be run in single-moment still, but I didn't see the need to add the flags to enable that option. I can change those.

I'll look at the changes for the reflectivity issue, too. Thanks for finding and pointing that out!

MicroTed committed:
11/24/2025:
     Changed tokens from from mp_nssl2m to mp_nssl
     Reconciled parallel changes in mpas_dmpar.F and mpas_atmphys_interface.F
     mpas_atmphys_init.F: Added flag f_qh for hail, but may be obsolete (f_qg is not used anymore)

MicroTed commented:
@jihyeonjang OK, I have fixed the Thompson issue and updated tokens from mp_nssl2m to mp_nssl (with a backward compatibility to allow 'mp_nssl2m' for config input). 
mpas_atmphys_init.F has a flag f_qg for graupel, and although it doesn't seem to do anything, I also added one for hail in case it is ever needed.

MicroTed committed:
mpas_atmphys_todynamics.F : For MYNN, apply tendency for 'nc' (droplet concentration) for mp_nssl

MicroTed commented:
One more commit fixes an oversight with the MYNN scheme where the cloud droplet concentration tendencies were not being applied but the mass tendency was. It would have some effect, but it's unclear how large.

MicroTed committed:
mpas_atmphys_driver_pbl.F : Separate f_nc from aerosol flags
  module_mp_thompson.F : fix print statement with correct file name

MicroTed committed:
mpas_atmphys_driver_microphysics.F : Fix an issue in the NSSL2M init call where gfortran debug compile causes an error.

MicroTed committed:
Merge branch 'develop' into dev-nssl3m

MicroTed commented on src/core_atmosphere/CMakeLists.txt:
```diff
@@ -355,6 +377,7 @@ add_library(core_atmosphere ${ATMOSPHERE_CORE_SOURCES}
             ${ATMOSPHERE_CORE_PHYSICS_MMM_SOURCES}
             ${ATMOSPHERE_CORE_PHYSICS_WRF_SOURCES}
             ${ATMOSPHERE_CORE_PHYSICS_NOAA_SOURCES}
+            ${ATMOSPHERE_CORE_PHYSICS_NSSL-MP_SOURCES}
```
I don't know cmake very well, but I see that other packages have explicit source file lists. I can add that for NSSLmp if that's needed.

mgduda commented on src/core_init_atmosphere/mpas_init_atm_cases.F:
```diff
@@ -6081,6 +6101,70 @@ subroutine init_atm_case_lbc(timestamp, block, mesh, nCells, nEdges, nVertLevels
+
+         ! check for missing moments that are needed
+         do k=1,nVertLevels
+            IF ( index_qs > 0 .and. index_ns > 0 ) THEN
```
Since `qs`, `ns`, etc. aren't defined as scalar constituents in the `Registry.xml` file, I think this code will lead to the use of unassociated pointers (`index_qs`, etc.).

Should there be additional changes in the `Registry.xml` file for the `init_atmosphere` core for scalar constituents beyond `qv`, `qc`, `qr`, `nifa`, and `nwfa`?

MicroTed replied:
Yes, there is work to be done here! We recently ran into an issue where gfortran thought a pointer was associated because it wasn't explicitly initialized as null. (in the GSL version) One question here is whether to add the microphysics scheme to the init namelist as in core_atmosphere and set the microphysics variables accordingly (i.e., have to know what scheme will be used). That would limit file sizes by not including unneeded variables. Or do we just provide all possible variables to have a generic IC/BC?

mgduda replied:
Given the number of new scalars, my inclination would be to only include these in LBC files if they will be needed in the model (i.e., if the NSSL scheme will be used). If we were to attach packages to each of the scalars, we could activate packages based on a specification in the `namelist.init_atmosphere` file of the microphysics scheme(s) that will be used in the model.

In future, though, it might be possible to scan through input files and activate packages for each scalar individually based on whether that scalar contains initial data in the input files. My thinking here is that, if we have no initial data for, say, `qs`, there would be no value in including it in the LBC files (as it would simply be zero everywhere, which is what the model will assume for a missing scalar anyway).

MicroTed replied:
Also, perhaps the diagnosed scalars (e.g., number concentration from a single-moment species) could be done in 'atmosphere' if the LBC is not found? For example, if there is snow mixing ratio, but the scheme needs the number concentration, it might be better to create that when mpas is reading lbcs and finds qs but not ns. It would be basically the same code that I have now on the init side, and would save having to write 'bogus' LBCs. That combined with your idea of checking what is even available would help limit the number of vars in the LBC file.

mgduda replied:
Apologies for the long delay in following up. I've finally spent a bit of time looking into the possibility of setting up packages for individual species based on either (1) the contents of the initial intermediate file or (2) a namelist-provided list of species. I've pushed a new branch, [demo/hydrometeor_packages](https://github.com/mgduda/MPAS-Model/commits/demo/hydrometeor_packages/), to my fork of the MPAS-Model repository with what I think might be a workable solution. In particular, the [last commit](https://github.com/mgduda/MPAS-Model/commit/4c81e2a4) on that branch adds the new package setup logic. Furthermore, I like your suggestion of diagnosing number concentrations and other scalars in the `atmosphere` core, as this would save on LBC file size.

If the approach take in the `demo/hydrometeor_packages` branch looks like it would work from your perspective, I think we could get that capability into the `develop` branch fairly quickly, and we could then rebase this PR branch and add packages for the new species in this PR (`qs`, etc.). Does that sound like a reasonable approach?

MicroTed replied:
Looks like a pretty clever approach to peek at the intermediate file. Sounds good to me.

As far as diagnosing LBC values, I guess that would need to be done in mpas_atm_update_bdy_tend?

mgduda replied:
I agree that `mpas_atm_update_bdy_tend` is probably the best place to diagnose number concentrations, etc. Around line 169 (in the v8.4.0 code) might be ideal, since we will have just read in the new boundary state into `lbc_*` time level 2 but not yet computed the time tendencies for the LBC fields: https://github.com/MPAS-Dev/MPAS-Model/blob/v8.4.0/src/core_atmosphere/dynamics/mpas_atm_boundaries.F#L169-L171 .

MicroTed replied:
Yunheng Wang wrote a subroutine to calculate reflectivity in the init phase for the initial output time, and that routine could be used as a template (or expanded) for diagnosing missing number concentrations, etc.  
https://github.com/NOAA-National-Severe-Storms-Laboratory/frdd-MPAS-Model/blob/da78534402b341adc4a921d062f1c2785ae35907/src/core_atmosphere/physics/mpas_atmphys_driver_microphysics.F#L1465 (corrected link)
It would be cleaner to call a similar routine than trying to handle various microphysics here. For NSSL-MP, it is ultimately calling the nssl_column_dbz and calcnfromq routines from the module.

MicroTed committed:
Added physics_noaa/NSSL-MP to Makefile include paths to fix errors with nvfortran (core_atmosphere)

mgduda commented:
@MicroTed Regarding this part of the PR description,

>Outstanding issue: At least in the physics_mmm directory, the .mod files are not updated by recompiling newer .F files. They have to be deleted in order to be replaced. I don't know if this is general or a gfortran quirk.

I think this should be fixed with the merge of https://github.com/NCAR/MMM-physics/pull/6 .

MicroTed commented:
> @MicroTed Regarding this part of the PR description,
> 
> > Outstanding issue: At least in the physics_mmm directory, the .mod files are not updated by recompiling newer .F files. They have to be deleted in order to be replaced. I don't know if this is general or a gfortran quirk.
> 
> I think this should be fixed with the merge of [NCAR/MMM-physics#6](https://github.com/NCAR/MMM-physics/pull/6) .

Oh, it has been so long... I recall somebody saying that this behavior was a 'feature' and that gfortran only generated a new module if the interface or variables had changed -- or I'm misremembering. I haven't had any problems with gnu-12.x (or 15.x)

MicroTed committed:
NSSL-MP: Update version of NSSl-MP (Externals.cfg) and modify code to pass the actual namelist
  filename to the init routine (to read the local nssl_mp_params namelist)
 Files modified: mpas_atm_core.F mpas_atmphys_driver_microphysics.F mpas_atmphys_init.F

MicroTed committed:
Removed docs/README.NSSLmp and added instead to physics_noaa/NSSL-MP

MicroTed committed:
Fix path errors to NSSL-MP in a couple Makefiles (no impact)
 Change commands in core_atmosphere/physics/Makefile for NSSL-MP to create symbolic links instead of copying, which avoids a needless recompile

mgduda commented:
@MicroTed We've just merged PR #1467 into the `develop` branch, so if you're able to merge `develop` to this PR branch, or to rebase this PR branch onto the HEAD of the `develop` branch, that should provide the groundwork needed for adding packages to the new scalars in ICs and LBCs that are introduced by this PR. I'm glad to help with any of this -- just let me know if you'd like me to do so (though I may need to be added as a GitHub 'collaborator' on your fork?).

MicroTed committed:
core_atmosphere/CMakeLists.txt : Comment out unneeded git calls

MicroTed committed:
Merge remote-tracking branch 'tedsrepo/develop' into dev-nssl3m

MicroTed commented:
@mgduda  I was out of the office last week, so I just got to this. It should be merged up to the current develop branch. I also found double-counting issue in the NSSL-MP that was causing excessive hail reflectivity in the 2-moment option. Hopefully good to go at this point.
I'm happy to add you as a collaborator, too.
